<launch>
  <arg name="model" default="QuadrotorQc11"/>
  <arg name="sim" value="0"/>
  <arg name="vicon" value="0"/>
  <arg name="vicon_fps" value="100"/>
  <arg name="mass" default="0.250"/>
  <arg name="odotopic" value="/$(arg model)/quadrotor_ukf/control_odom"/>

  <group ns="$(arg model)">

    <param name="mass" value="$(arg mass)"/>

    <node pkg="master_discovery_fkie" type="master_discovery" name="master_discovery">
      <param name="mcast_group" value="224.0.0.1" />
    </node>

    <node pkg="master_sync_fkie" type="master_sync" name="multimaster_client_sync">
      <rosparam file="$(find state_control_vehicle)/config/server_multimaster.yaml" />
    </node>

    <node pkg="nodelet"
      type="nodelet"
      name="standalone_nodelet"
      args="manager"
      output="screen">
      <rosparam file="$(find mav_manager)/config/mav_manager_params.yaml" />
    </node>

    <node pkg="nodelet"
      type="nodelet"
      name="trackers_manager"
      args="load trackers_manager/TrackersManager standalone_nodelet"
      output="screen">
      <rosparam file="$(find trackers_manager)/config/trackers.yaml"/>
      <rosparam file="$(find state_control_vehicle)/config/QuadrotorQ2_gains.yaml"/>
      <remap from="~odom" to="$(arg odotopic)"/>
      <remap from="~cmd" to="position_cmd"/>
    </node>

    <node pkg="nodelet"
      type="nodelet"
      args="load so3_control/SO3ControlNodelet standalone_nodelet"
      name="so3_control"
      required="true"
      output="screen">
      <rosparam file="$(find state_control_vehicle)/config/QuadrotorQ2_gains.yaml"/>
      <param name="mass" value="$(arg mass)"/>
      <param name="use_external_yaw" value="false"/>
      <remap from="~odom" to="$(arg odotopic)"/>
      <remap from="~position_cmd" to="position_cmd"/>
      <remap from="~so3_cmd" to="so3_cmd"/>
      <remap from="~motors" to="motors"/>
    </node>

    <node pkg="nodelet"
      type="nodelet"
      name="quad_encode_msg"
      args="load quad_encode_msg/QuadEncodeMsg standalone_nodelet"
      output="screen"
      unless="$(arg sim)">
      <remap from="~so3_cmd" to="so3_cmd"/>
      <remap from="~serial_msg" to="to_robot"/>
    </node>

    <node pkg="mav_manager"
      type="mav_services"
      name="mav_services"
      clear_params="true"
      output="screen">
      <rosparam file="$(find state_control_vehicle)/config/mav_manager_params.yaml" />
      <remap from="~odom" to="$(arg odotopic)"/>
      <remap from="odom" to="$(arg odotopic)"/>
      <remap from="~des_odom" to="des_odom"/>
    </node>

    <node pkg="nodelet"
      type="nodelet"
      args="standalone qualcomm_interface/SO3CmdToQualcomm"
      name="so3cmd_to_qualcomm_nodelet"
      required="true"
      clear_params="true"
      output="screen">
      <remap from="~odom" to="$(arg odotopic)"/>
      <remap from="~so3_cmd" to="so3_cmd"/>
    </node>

    <node pkg="vio_qc"
      type="MVSampleVISLAMEagle"
      name="MVSampleVISLAMEagle"
      output="screen" args = "-a">
      <remap from="~vio_pose" to="vio_pose"/>
      <remap from="~imu" to="imu"/>
    </node>

    <node pkg="quadrotor_ukf"
      type="quadrotor_ukf"
      name="quadrotor_ukf"
      output="screen">
      <remap from="~odom_slam" to="vio_pose"/>
      <remap from="~imu" to="imu"/>
    </node>

    <node pkg="topic_tools"
      type="relay"
      name="over_and_over_again"
      args="$(arg odotopic) odom">
    </node>

  </group>
</launch>
